src/domains/access.ml

```diff
@@ -455,23 +455,30 @@ let check_accs (prev_r,prev_lp,prev_w) (conf,w,loc,e,lp) =
     (new_r, new_lp, new_w)
   | _ -> (prev_r,prev_lp,prev_w)
 
-let check_safe ls (accs,lp) prev_safe =
+(* Check if two accesses race and if yes with which confidence *)
+let conflict2 (conf,mhp,w,loc,e,lp) (conf2,mhp2,w2,loc2,e2,lp2) =
+  if (not w) && (not w2) then
+    None (* two read/read accesses do not conflict *)
+  else if not (LSSet.is_empty @@ LSSet.inter lp lp2) then
+    None (* the labelled string set excludes that these conflict *)
+  else if not (MHP.may_happen_in_parallel mhp mhp2) then
+    None (* They may not be in parallel *)
+  else
+    Some (max conf conf2)
+
+let check_safe ls (accs,_) prev_safe =
   if ls = None then
     prev_safe
   else
-    let ord_enum = Set.backwards accs in (* hope that it is not nil *)
-    let lp_start = (fun (_,_,_,_,lp) -> lp) (BatOption.get (BatEnum.peek ord_enum)) in
-    (* ignore(printf "starting with lockset %a\n" LSSet.pretty lp_start); *)
-    match BatEnum.fold check_accs (None, lp_start, false) (Set.backwards accs), prev_safe with
-    | (None, _,_), _ ->
-      (* ignore(printf "this batch is safe\n"); *)
+    let cart = BatSet.cartesian_product accs accs in
```

Comment:
I think this cartesian product currently includes unnecessarily many pairs, even though by symmetry only half should be needed. Also pairs with itself should be unnecessary, at least they were never checked in the old linear fold code.

Reply:
Also, I'm a bit confused why this cartesian product is here, because below `check_safe` is still folded linearly over some set of accesses. Isn't this now doing O(n^2) work _for each access_, totalling O(n^3) overall if accounting to that outer fold?

Reply:
The way I understood it, the fold is over the different regions (or partitions) and the pair-wise race check than happens inside such a region. So if we have `m` regions of `n` accesses each, we'd have `O(m*n^2)`, right?

Reply:
> Also pairs with itself should be unnecessary

I don't think so. Consider e.g. a global that is accessed only from one(!) thread with a non-unique thread id.

> at least they were never checked in the old linear fold code.

Reply:
> This is true, but I am not so sure how to elegantly do such a half-product (for lack of a better word).

If the elements have a `compare` then one could just filter the pairs with `fun (x, y) -> E.compare x y <= 0` or something. I think there are other places where I've done it like that.
Of course that's not optimal either because it involves first constructing all the pairs, but at least it avoids duplicating the possibly more expensive checks and warning generations.

Reply:
> Consider e.g. a global that is accessed only from one(!) thread with a non-unique thread id.

You're probably right about that. Since we're going to make it more pairwise anyway, I cannot be bothered to figure out exactly how that worked out before. Probably just because such accesses would have no locks nor thread ID, so some set is empty and hence trivially racing.